example/src/MutableReference/Prop.hs

```diff
@@ -169,23 +176,29 @@ prop_shrinkParallelMinimal = shrinkPropertyHelper (prop_referencesParallel RaceC
       , Internal (Inc   ref0)   (Symbolic var1)
       ]
 
-instance Read (Reference Symbolic (Opaque (IORef Int))) where
-  readPrec     = Reference . Symbolic <$> readPrec
-  readListPrec = readListPrecDefault
-
-instance Read (Internal Action) where
+instance Read (Untyped Action) where
 
-  readPrec = choice
-    [ Internal <$> (New   <$ lift (string "New"))  <*> readPrec
-    , Internal <$> (Read  <$ lift (string "Read")  <*> readPrec) <*> readPrec
-    , Internal <$> (Write <$ lift (string "Write") <*> readPrec <*> readPrec) <*> readPrec
-    , Internal <$> (Inc   <$ lift (string "Inc")   <*> readPrec) <*> readPrec
-    ]
+  readPrec = parens $ do
+    Ident i <- lexP
+    case i of
+      "New"   -> return (Untyped New)
+      "Read"  -> u (Read  <$> readPrec)
+      "Write" -> u (Write <$> readPrec <*> readPrec)
+      "Inc"   -> u (Inc   <$> readPrec)
+    where
+      u :: (Show a, Typeable a)
+        => ReadPrec (Action Symbolic a) -> ReadPrec (Untyped Action)
+      u x = Untyped <$> prec appPrec (step x)
+      appPrec = 10
 
-  readListPrec = readListPrecDefault
 
 deriving instance Eq1 v => Eq (Action v resp)
-
-instance Eq (Internal Action) where
-  Internal act1 sym1 == Internal act2 sym2 =
-    cast act1 == Just act2 && cast sym1 == Just sym2
+instance Eq (Untyped Action) where
```

Comment:
I'm a bit confused here. Why do we need equality for untyped actions? I thought we needed equality of programs (which in turn consist of internal actions) for checking if counterexamples are minimal?

Reply:
Equality of `Internal Action` now defers to equality of `Untyped Action` to compare the first field (which I realize could just use a cast here). The pattern I'm going for here is that the user only needs to derive instances for either `Action` or `Untyped Action`.

Reply:
Ah that makes sense. Deriving equality for `Action` and then defining it for `Untyped Action` using cast would be nice. (And then we could template Haskell away that later.)